### ISSUE TITLE
Up the minimum version of PHPCS to 3.7.0 as soon as it is released

### ISSUE THREAD

**Issue by jrfnl** [WordPress/WordPress-Coding-Standards#2048]
## Introduction

The last three PHP releases, PHP 7.4, 8.0 and 8.1, introduced more syntax changes and new keywords in PHP than basically all PHP versions between PHP 5.0 to PHP 7.3 combined.

The work on getting all of these (tokenizer) changes accounted for and backfilled in PHPCS to allow for running PHPCS on older PHP versions, but still get the same token stream, is still on-going.

Along the same lines, the work of adding support for these new syntaxes and adding utility methods to PHPCSUtils to allow supporting these new syntaxes on older PHPCS versions as well is also still on-going.

### PHPCS Runtime compatibility

In this context, runtime compatibility means that PHPCS will _run_ on a PHP version without fatal errors or deprecation notices.

PHP | PHPCS
--- | ---
7.4 | 3.4.2 and higher
8.0 | 3.5.7 and higher
8.1 | 3.6.2 and higher

### Syntax support

In this context, syntax support means that the Tokenizer in PHPCS supports all new syntaxes of a new PHP version. It does not imply that all sniffs have yet been adjusted to allow for these new syntaxes correctly, nor that all utility functions included in PHPCS itself support the new syntaxes yet.

There may also still be minor bugs in the Tokenizer support for these new syntaxes. The PHPCS version listed is the first version in which all new syntaxes are supported and **_major_** known bugs have been resolved.

PHP | PHPCS
--- | ---
7.4 | 3.5.6 and higher
8.0 | 3.6.1 and higher
8.1 | 3.7.0 and higher, though this may still be partial support (expected)

## The problem

For sniffs to stay accurate, they need to be updated to take these new syntaxes into account.

For example, PHP 7.4 introduced arrow functions.
This means that all sniffs looking explicitly for function declarations - for instance to verify the formatting of these - will need to be updated to now also include arrow functions.

At the same time, sniffs which need to _skip over_ function declarations to prevent false positives and such, will now also need to _skip over_ arrow functions.

Arrow functions are supported in PHPCS as of PHPCS 3.5.3 (very buggy), but realistically PHPCS 3.5.6 is needed as by that time the worst bugs in the Tokenizer arrow function support were resolved.

While PHPCSUtils can - up to a point - provide utilities to backport support for these kind of new syntaxes, there are a couple of limitations:
1. Backported support via PHPCSUtils will **always** create overhead in a sniff while processing code.
2. Using backported support via PHPCSUtils creates mental overhead for contributors as they need to be intimately familiar with all related tokenizer issues to take into account and how.
3. There are syntaxes for which support simply _cannot_ be backported.

Take for example: PHP 8.0 attributes. In PHP < 8.0, these are tokenized as comments.

Now, while that isn't that much of a problem if the attribute is on a line by itself, this is problematic when the attribute is combined with code on the same line **_or_** when multi-line attributes are used.

---
<details>
  <summary>Some examples of the problems with attributes</summary>

```php
#[SingleLineAttribute]
function foobar() {}
```

With the above example, PHP < 8.0 on PHPCS < 3.6.0 will tokenize the whole attribute line as a comment.

```php
function foobar(
    #[ParameterAttribute] int|string $param
) {}
```

With this example, PHP < 8.0 on PHPCS < 3.6.0 will tokenize the whole _line_ as a comment, that is _including the type declaration and the parameter variable.

```php
function foobar(
    #[MultilineAttribute(
        10,
        []
    )]
    int|string $param
) {}
```

And with this example, the first line of the attribute will be tokenized as a comment on PHP < 8.0 on PHPCS < 3.6.0, while the next few lines will be tokenized normally, but will confuse the setting of conditions/nested parentheses etc.

</details>

---

There is no way for PHPCSUtils to correct the token stream and for PHP < 8.0, the token stream will be very wrong unless PHPCS 3.6.0 or higher is used and for correct helper function handling, PHPCS 3.6.1 is needed.

## Recommendation

All in all, I would like to **strongly** recommend to up the minimum supported PHPCS version to PHPCS 3.7.0 as soon as it comes out and before the release of WPCS 3.0.0.

PHPCS 3.7.0 was expected a few weeks ago and while there is no telling when exactly it will land, it is likely to be released some time over the next month or so.

While there will probably still be upstream sniffs which need fixing for the new syntaxes, using PHPCS 3.7.0 as the new minimum for WPCS will save us a lot of headaches when adding support for the new syntaxes to the WPCS sniffs.

## Action Checklist

- [ ] Update the `composer.json` file as soon as PHPCS 3.7.0 has been released to set the new minimum version
